server/src/room/id.rs

```diff
@@ -0,0 +1,117 @@
+use once_cell::sync::Lazy;
+use std::error::Error;
+use std::fmt::{self, Write};
+use std::str::FromStr;
+
+// Alphanumerics, excluding iIoO01, since they look too similar.
+static ROOM_ID_CHARS: Lazy<Vec<char>> = Lazy::new(|| {
+    "23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghjklmnpqrstuvwxyz"
+        .chars()
+        .collect()
+});
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub struct RoomId(u128);
+
+impl RoomId {
+    pub fn random() -> RoomId {
+        RoomId(rand::random())
+    }
+}
+
+impl fmt::Display for RoomId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // use a prefix ("r") to allow us to detect possible future changes to this format
+        f.write_char('r')?;
+        let len: u128 = ROOM_ID_CHARS.len() as u128;
+        let mut rest = self.0;
+        while rest > 0 {
+            f.write_char(ROOM_ID_CHARS[(rest % len) as usize])?;
+            rest /= len;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for RoomId {
+    type Err = InvalidRoomIdError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.chars();
+        match iter.next() {
+            Some('r') => {}
+            Some(_) | None => {
+                return Err(InvalidRoomIdError(s.to_owned()));
+            }
+        }
+        let mut coefficient: u128 = 1;
+        let mut result: u128 = 0;
+        for ch in iter {
+            match ROOM_ID_CHARS.binary_search(&ch) {
+                Ok(idx) => {
+                    result += idx as u128 * coefficient;
+                    coefficient *= ROOM_ID_CHARS.len() as u128;
```

Comment:
This can cause an overflow which will panic on debug builds (it silently overflows on release builds, which is okay in this case).

I need to prevent this multiply if we've just handled the last character (happy path), and I need to return an `InvalidRoomIdError` if the room id is too long.